### DISCUSSION TITLE
Getting bunch of "None"s when extracting tables using pdfplumber

### DISCUSSION THREAD

**Discussion by domdrag** [jsvine/pdfplumber#719] [Ask for help with specific PDFs]
Hello, hopefully, I'm in the right place. I'm having troubles using pdfplumber when extracting tables.
My guess is that I'm doing something wrong and I would really appreciate if someone could guide me.

So basically, here's the pdf I'm extracting tables from
[pdf.pdf](https://github.com/jsvine/pdfplumber/files/9439561/pdf.pdf).
I want to extract all rows in the tables. I'm not interested in the titles, headers and all of that, I just need rows.
I'm able to extract all rows except for one shown in this pic:
![issue](https://user-images.githubusercontent.com/58782481/187072635-50b4de7c-bccd-4bf4-87c6-7d6f678d29b4.png)

For some reason I'm getting the following output:
```python
...
['101', '07.13', 'PTD', '04:42', '11:27', '7,92', '1,30', '', '401', '07.13', 'Dubrava \nsmjer Savski gaj', '11:27', '18:00', '7,72', '4,00', '701', '', '', '', '', '', '', '']
['102', None, None, None, None, None, None, None, '402', None, None, None, None, None, None, '702', None, None, None, None, None, None, None]
['103', '07.15', 'PTD', '04:51', '11:07', '7,43', '1,15', '', '403', '07.15', 'Dubrava \nsmjer Dubec', '11:07', '17:36', '7,65', '3,60', '703', '07.15', 'Dubrava \nsmjer Dubec', '17:36', '00:17', '7,85', '2,28', '4,40']
...
```

Here's the code I'm using:
```python
import pdfplumber

def readAllServices():
    PDF = pdfplumber.open('pdf.pdf')
    for page in PDF.pages:
        tables = page.dedupe_chars().find_tables()
        for tableId in tables:
            table = tableId.extract()
            for serviceLine in table:
                if(isinstance(serviceLine[0], str) and serviceLine[0].isnumeric()):
                    print(serviceLine)

readAllServices()
```
Can someone help me here? What am I missing? I've taken a look at this forum and documentation, but haven't found anything helpful. Thanks in advance.

EDIT: I've just noticed that I don't need a loop through pages, that part left from the original code that is much bigger. So this is almost a minimal reproducible example. But anyway, that shouldn't matter for my issue.

**Comment by jsvine** [comment]
Hi @domdrag, and thanks for your interest in this library. When I encounter situations like these, I like to use `pdfplumber`'s visual debugging tools:

```python
import pdfplumber
pdf = pdfplumber.open("discussion-719.pdf")
page = pdf.pages[0]
im = page.to_image(resolution=150)
im.reset().debug_tablefinder()
```

The result looks like this:

![image](https://user-images.githubusercontent.com/534702/188740718-46d3c87f-1dbc-4772-9d87-9f795ad216b2.png)

Looking at the row you had difficulty with, it seems there's a random extra set of lines (or, more accurately, edges of a `rect`) in the middle of the row, fooling `pdfplumber`:

<img width="809" alt="Screen Shot 2022-09-06 at 5 18 34 PM" src="https://user-images.githubusercontent.com/534702/188740045-d8401f5c-6d3f-41b0-8952-65cd7ad459af.png">

Poking around a bit more, I'm noticing that the offending `rect` objects all have `"non_stroking_color" == (0, 0, 1)`:

```python
(
    im.reset()
    .draw_rects(
        [ r for r in page.rects if r["non_stroking_color"] == (0, 0, 1) ],
        stroke_width=5
    )
)
```

![image](https://user-images.githubusercontent.com/534702/188742951-513f5430-9061-4af1-86eb-4da14c68df1a.png)

It seems that, for whatever reason, the program that drew the table decided to color in the blue (`(0, 0, 1)`) columns in multiple segments sometimes. We can remove those rects using `page.filter(...)`, and try again:

```
filtered = (
    page
    .filter(lambda obj: not (
        obj.get("non_stroking_color") == (0, 0, 1) 
        # The following line makes sure we're only rejecting the narrow blue rects
        and obj.get("width", 100) < 100 
    ))
)
```

Visually debugging again:

```python
filtered.to_image().debug_tablefinder()
```

... produces this:

![image](https://user-images.githubusercontent.com/534702/188745242-783ea4bb-b1a7-4e9c-8107-a2294496e92e.png)

... which seems right. And this:

```python
filtered.extract_tables()[0][15]
```

... produces *almost* what you're looking for:

```
['110022',
 '0077..1144',
 'PPTTDD            \nPTD',
 '0066::4455    \n12:57',
 '00    99::3366      \n17:29',
 '88,,6600',
 '',
 '33,,4488',
 '440022',
 '',
 '',
 '',
 '',
 '',
 '',
 '770022',
 '0077..1144',
 'DDuubbrraavvaa  \nsmjer Dubec',
 '1177::2299',
 '0000::0066',
 '77,,7788',
 '22,,1100',
 '44,,5522']
```

... but which reveals another issue, which is that the PDF contains some duplicate characters (cf. https://github.com/jsvine/pdfplumber/issues/71) in that row. So, the final touch:

```
filtered = (
    page
    .filter(lambda obj: not (
        obj.get("non_stroking_color") == (0, 0, 1) 
        and obj.get("width", 100) < 100
    ))
    .dedupe_chars() # <<< 
)
```

... so that `filtered.extract_tables()[0][15]` gets you:

```
['102',
 '07.14',
 'PTD      \nPTD',
 '06:45  \n12:57',
 '0  9:36   \n17:29',
 '8,60',
 '',
 '3,48',
 '402',
 '',
 '',
 '',
 '',
 '',
 '',
 '702',
 '07.14',
 'Dubrava \nsmjer Dubec',
 '17:29',
 '00:06',
 '7,78',
 '2,10',
 '4,52']
```

**Reply by domdrag** [reply]
Hi @jsvine, thank you so much for answering. This looks great! I wouldn't have solved this in million years. And great work with the library!
